script/shared/agent.ts

```diff
@@ -0,0 +1,147 @@
+import path from "path";
+import fs from "fs";
+
+import { Config } from "./config";
+import { defaultContextPath } from "./constants";
+import { Context, ContextHook } from "./context";
+import { extractByte32AddrFromBech32 } from "./utils";
+import { getContractInfo } from "./wasm";
+
+export type HplAgentConfig = {
+  name: string;
+  domainId: string;
+  chainId: string;
+  protocol: "cosmos";
+  rpcUrls: { http: string }[];
+  grpcUrl: string;
+  canonicalAsset: string; // usually same as gas token
+  bech32Prefix: string; // hrp
+  gasPrice: {
+    amount: string;
+    denom: string;
+  };
+  contractAddressBytes: 32;
+  index: {
+    from: number;
+    chunk: number;
+  };
+  blocks: {
+    reorgPeriod: 0; // instant finality ⭐️
+  };
+
+  mailbox: string; // hexed
+  interchainGasPaymaster: string; // hexed
+  validatorAnnounce: string; // hexed
+  merkleTreeHook: string; // hexed
+  testRecipient: string; // hexed
+};
+
+export async function fromContext(
+  network: Config["networks"][number],
+  context: Context
+): Promise<HplAgentConfig> {
+  const toHex = (v: string) => `0x${extractByte32AddrFromBech32(v)}`;
+
+  const { hooks, core, test } = context.deployments;
+
+  const mailboxAddr = core?.mailbox?.address!;
+  const mailboxContractInfo = await getContractInfo(network, mailboxAddr);
+
+  const igp =
+    findHook(hooks?.default!, "hpl_igp") ||
+    findHook(hooks?.required!, "hpl_igp");
+  if (!igp) throw new Error("no igp on this context");
+
+  const merkleTreeHook =
+    findHook(hooks?.default!, "hpl_hook_merkle") ||
+    findHook(hooks?.required!, "hpl_hook_merkle");
+  if (!merkleTreeHook) throw new Error("no merkle tree hook on this context");
+
+  const agent: HplAgentConfig = {
+    name: network.id.split("-").join(""),
+    domainId: network.domain.toString(),
+    chainId: network.id,
+    protocol: "cosmos",
+
+    rpcUrls: [{ http: network.endpoint.rpc }],
+    grpcUrl: network.endpoint.grpc,
+    canonicalAsset: network.gas.denom,
+    bech32Prefix: network.hrp,
+
+    gasPrice: {
+      amount: network.gas.price,
+      denom: network.gas.denom,
+    },
+    contractAddressBytes: 32,
+
+    index: {
+      from:
+        // sub 1 block to make sure we don't miss any block
+        parseInt(mailboxContractInfo.contract_info.created.block_height) - 1,
+      chunk: 10_000,
+    },
+    blocks: {
+      reorgPeriod: 0,
+    },
+
+    // contract addresses
+    mailbox: toHex(core?.mailbox?.address!),
+    validatorAnnounce: toHex(core?.validator_announce?.address!),
+    interchainGasPaymaster: toHex(igp.address),
+    merkleTreeHook: toHex(merkleTreeHook.address),
+    testRecipient: toHex(test?.msg_receiver?.address!),
+  };
+
+  return agent;
```

Comment:
Consider running this through our zod type to make sure its well formatted https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/f2a0e92c7ff35ae0fef17d10ec1e2c2713578d34/typescript/sdk/src/metadata/agentConfig.ts#L211

Reply:
Gotcha, I'll make this happen following work.